src/Microsoft.Azure.Batch.SoftwareEntitlement.Server/Controllers/SoftwareEntitlementsController.cs

```diff
@@ -57,43 +57,27 @@ public SoftwareEntitlementsController(ServerOptions serverOptions, ILogger logge
         [HttpPost]
         [Produces("application/json")]
         public IActionResult RequestEntitlement(
-            [FromBody] SoftwareEntitlementRequest entitlementRequest,
+            [FromBody] SoftwareEntitlementRequestBody entitlementRequestBody,
             [FromQuery(Name = "api-version")] string apiVersion)
         {
             try
             {
-                if (!IsValidApiVersion(apiVersion))
+                var (parameters, requestError) = TryExtractParameters(apiVersion, entitlementRequestBody);
+                if (requestError != null)
                 {
-                    return CreateInvalidApiVersionError(entitlementRequest, apiVersion);
+                    return CreateBadRequestResponse(requestError);
                 }
 
-                _logger.LogInformation(
-                    "Selected api-version is {ApiVersion}",
-                    apiVersion);
-
-                if (entitlementRequest == null)
-                {
-                    return CreateMissingRequestError();
-                }
-
-                _logger.LogInformation(
-                    "Requesting entitlement for {Application}",
-                    entitlementRequest.ApplicationId);
-                _logger.LogDebug("Request token: {Token}", entitlementRequest.Token);
-
-                var remoteAddress = HttpContext.Connection.RemoteIpAddress;
-                _logger.LogDebug("Remote Address: {Address}", remoteAddress);
-
                 Errorable<NodeEntitlements> verificationResult = _verifier.Verify(
-                    entitlementRequest.Token,
+                    parameters.Token,
                     _serverOptions.Audience,
                     _serverOptions.Issuer,
-                    entitlementRequest.ApplicationId,
-                    remoteAddress);
+                    parameters.ApplicationId,
+                    parameters.RemoteAddress);
```

Comment:
Looking at this method alone, it implies that `RemoteAddress` is somehow extracted from `entitlementRequestBody`, which is misleading, given that it's extracted from ambient state.

Reply:
True. Perhaps the HTTP context (or connection, or even just the IP address itself) could be passed in to the `TryExtractParameters` method. Any preference?

Reply:
I'm getting hung up on the method naming - the origin IP address of the connection isn't a parameter that gets extracted from the request body, it comes from somewhere else. If the method is called `TryExtractParameters` then it should do what it says on the tin, nothing more.

Reply:
But if (say) an `HttpContext` was passed in to the method, the IP address _would_ be extracted from that.

In a subsequent change, though, I'm thinking of building a (logical) `EntitlementVerificationRequest` in that method, to be passed directly to the `SoftwareEntitlement` layer. If/when I get around to doing that, the method will need to be renamed anyway, maybe to something like `BuildVerificationRequest()` (?). Unless you've got a great idea for what to call it just now, could we save that discussion until then?